### DISCUSSION TITLE
3 features for FileUpload

### DISCUSSION THREAD

**Discussion by bilogic** [filamentphp/filament#8148] [Ideas]
I have been messing around with FileUpload and SpatieMediaLibraryFileUpload these few days, here is what I concluded

1. Moving from single to multiple files and vice versa is quite tedious: need to cast the field as an array + convert existing string value to JSON, why not use JSON by default even for single file? This will make switching between single/multiple files more seamless. Change is a constant, there were many times where it was painful enough that we just said, forget it, you can only upload 1 file

2. Since multitenancy is supported by filament, why not have a built-in way to prevent filename clashes when different tenants upload files with the same names? Maybe storing them in separate folders or something

3. It could be argued that we can just use SpatieMediaLibraryImageColumn, but it cannot be used when there are no models

Thank you

**Comment by danharrin** [comment]
> Moving from single to multiple files and vice versa is quite tedious

I don't understand the problem really, you can use `multiple()->maxFiles(1)` if you want to achieve that data structure

> Since multitenancy is supported by filament, why not have a built-in way to prevent filename clashes when different tenants upload files with the same names

The default behaviour generates random file names that will not clash, so again I don't understand the problem. You can specify a directory if you want `->directory(Filament::getTenant()->id)`

> It could be argued that we can just use SpatieMediaLibraryImageColumn, but it cannot be used when there are no models

Right... because Spatie forces you to associate files with models, they store references to files in a common `media` table and no references on the model itself, hence why that is required. Again, I don't understand

**Reply by bilogic** [reply]
Thanks for getting back.

1. `multiple()->maxFiles(1)` this solves my first point
2. Based on my testing of `FileUpload`, if I add the `->preserveFilenames()` option, that would lead to clashes correct? What I found is that most users prefer to retain the original filename especially for non-images. `->directory()` would help with this problem, but it would not be obvious to a first time filament user and would have to be introduced to every single `FileUpload`. What I'm trying to suggest is, clash free right out of the box even with options enabled
3. My last point was made because often in other repos, there would be suggestions to use an alterative (in this case `SpatieMediaLibraryImageColumn`) to achieve points 1 and 2, but that would introduce new restrictions (i.e. models are required)

**Reply by danharrin** [reply]
2) The form builder is separate from the panel builder and it doesn't understand that the tenant even exists. I don't think passing the tenant ID to `directory()` is ugly enough to warrant further abstraction, especially since `preserveFilenames()` is not enabled by default. You could even use `storeFileNamesIn()` to achieve the same result without clashes.
